ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```
@@ -0,0 +1,777 @@
+      if (operationException.get() != null || numChunksWrittenOut.get() == numChunksTotal) {
+        completeRead();
+        operationCompleted = true;
+      }
+    }
+  }
+
+  /**
+   * Complete the read from this {@link ReadableStreamChannel} by invoking the callback and marking the future.
+   */
+  void completeRead() {
+    if (readIntoCallbackCalled.compareAndSet(false, true)) {
+      if (operationException.get() != null) {
+        bytesWritten = null;
+      }
+      readIntoFuture.done(bytesWritten, operationException.get());
+      if (readIntoCallback != null) {
+        readIntoCallback.onCompletion(bytesWritten, operationException.get());
+      }
+    }
+  }
+
+  /**
+   * As part of the poll, GetBlobOperation fetches new requests to be issued and updates the state based on
+   * whether previously issued requests have timed out and so on. Additionally, any writes of previously retrieved
+   * chunk buffers into the {@link AsyncWritableChannel} passed in by the caller also happens within this method.
+   * @param requestRegistrationCallback the {@link RequestRegistrationCallback} to call for every request that gets
+   *                                    created as part of this poll operation.
+   */
+  @Override
+  void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
+    if (operationCompleted) {
+      return;
+    }
+    if (firstChunk.isReady() || firstChunk.isInProgress()) {
+      firstChunk.poll(requestRegistrationCallback);
+    }
+    if (firstChunk.isComplete()) {
+      // Although an attempt is made to write to the channel as soon as a chunk is successfully retrieved,
+      // the caller might not have called readInto() and passed in a channel at the time. So an attempt is always
+      // made from within poll.
+      maybeWriteToChannel();
+      // If this is a composite blob, poll for requests for subsequent chunks.
+      if (dataChunks != null) {
+        for (GetChunk dataChunk : dataChunks) {
```

should we fail fast to check for operationException at places like here. Since an exception can be thrown via chunkAsyncWriteCallback, thought no point in iterating over the dataChunks.